_posts/2016-09-30-subtype-typeclasses.md

```diff
@@ -0,0 +1,402 @@
+---
+layout: post
+title: Subtype type classes don't work
+
+meta:
+  nav: blog
+  author: adelbertc
+  pygments: true
+---
+
+## Problem
+
+The common encoding of type classes in Scala relies on subtyping. This singular
+fact gives us a certain cleanliness in the code, but at what cost?
+
+Consider the following hierarchy of type classes. A similar hierarchy can be
+found in both [Cats][cats] and [Scalaz 7][scalaz7].
+
+```scala
+trait Functor[F[_]]
+
+trait Applicative[F[_]] extends Functor[F]
+
+trait Monad[F[_]] extends Applicative[F]
+
+trait Traverse[F[_]] extends Functor[F]
+```
+
+For purposes of demonstration I will be using Cats for the rest of this post,
+but the same arguments apply to Scalaz 7.
+
+We will also assume that there is syntax accompanying this hierarchy, allowing
+us to call methods like `map`, `flatMap`, and `traverse` directly on some
+`F[A]`, provided `F` has the appropriate type class instances (`Functor`,
+`Monad`, and `Traverse`, respectively).
+
+```scala
+import cats._
+import cats.implicits._
+```
+
+One important consequence is we can use for comprehensions in methods
+parameterized over some `Monad`.
+
+```scala
+def foo[F[_]: Monad]: F[Int] = for {
+  a <- Monad[F].pure(10)
+  b <- Monad[F].pure(20)
+} yield a + b
+```
+
+Notice that due to how for comprehensions [desugar][forcomp], there is also
+a call to `map` in there. Since our type class hierarchy is encoded via
+subtyping Scala knows a `Monad[F]` implies a `Functor[F]`, so all is well.
+Or is it?
+
+Consider a case where we want to abstract over a data type that has
+both `Monad` and `Traverse`.
+
+```scala
+// Ignore the fact we're not even using `Traverse` - we can't even call `map`!
+def foo[F[_]: Monad: Traverse]: F[Int] = Monad[F].pure(10).map(identity)
+// <console>:19: error: value map is not a member of type parameter F[Int]
+//        def foo[F[_]: Monad: Traverse]: F[Int] = Monad[F].pure(10).map(identity)
+//                                                                   ^
+// <console>:19: error: missing argument list for method identity in object Predef
+// Unapplied methods are only converted to functions when a function type is expected.
+// You can make this conversion explicit by writing `identity _` or `identity(_)` instead of `identity`.
+//        def foo[F[_]: Monad: Traverse]: F[Int] = Monad[F].pure(10).map(identity)
+//                                                                       ^
+```
+
+We're already in trouble. In order to call `map` we need `F` to have a
+`Functor` instance, which it does via `Monad` as before.. but now also via
+`Traverse`. It is for precisely this reason that this does not work. Because
+our encoding of type classes uses subtyping, a `Monad[F]` **is a** `Functor[F]`.
```

Comment:
Ah yes that's a good talk to link to. I think such a topic warrants it's own blog post, so I think I'll just link to the talk for now. Thanks for the reminder!